src/main/network/relay/token_bucket.rs

```diff
@@ -72,10 +78,42 @@ impl TokenBucket {
         self.balance = self
             .balance
             .checked_sub(decrement)
-            .ok_or(next_refill_span)?;
+            .ok_or_else(|| self.compute_conforming_duration(decrement, next_refill_span))?;
         Ok(self.balance)
     }
 
+    /// Computes the duration required to refill enough tokens such that our
+    /// balance can be decremented by the given `decrement`. `next_refill_span`
+    /// is the duration until the next refill, which may be less than a full
+    /// refill interval.
+    fn compute_conforming_duration(
+        &self,
+        decrement: u64,
+        next_refill_span: SimulationTime,
+    ) -> SimulationTime {
+        let required_token_increment = decrement.saturating_sub(self.balance);
+
+        let num_required_refills = {
+            // Same as `required_token_increment.div_ceil(self.refill_increment);`
+            let num_refills = required_token_increment / self.refill_increment;
+            let remainder = required_token_increment % self.refill_increment;
+            if remainder > 0 && self.refill_increment > 0 {
+                num_refills + 1
+            } else {
+                num_refills
+            }
+        };
+
+        match num_required_refills {
+            0 => SimulationTime::ZERO,
+            1 => next_refill_span,
+            _ => next_refill_span.saturating_add(
+                self.refill_interval
+                    .saturating_mul(num_required_refills.saturating_sub(1)),
```

Comment:
nit: maybe `num_required_refills.checked_sub(1).unwrap()`, since if this fails, then something else went wrong